File: elixir/protein-translation/protein_translation_test.exs
if !System.get_env("EXERCISM_TEST_EXAMPLES") do
  Code.load_file("protein_translation.exs", __DIR__)
end

ExUnit.start()
ExUnit.configure(exclude: :pending, trace: true)

defmodule ProteinTranslationTest do
  use ExUnit.Case

  test "AUG translates to methionine" do
    assert ProteinTranslation.of_codon("AUG") == {:ok, "Methionine"}
  end

  test "identifies Phenylalanine codons" do
    assert ProteinTranslation.of_codon("UUU") == {:ok, "Phenylalanine"}
    assert ProteinTranslation.of_codon("UUC") == {:ok, "Phenylalanine"}
  end

  test "identifies Leucine codons" do
    assert ProteinTranslation.of_codon("UUA") == {:ok, "Leucine"}
    assert ProteinTranslation.of_codon("UUG") == {:ok, "Leucine"}
  end

  test "identifies Serine codons" do
    assert ProteinTranslation.of_codon("UCU") == {:ok, "Serine"}
    assert ProteinTranslation.of_codon("UCC") == {:ok, "Serine"}
    assert ProteinTranslation.of_codon("UCA") == {:ok, "Serine"}
    assert ProteinTranslation.of_codon("UCG") == {:ok, "Serine"}
  end

  test "identifies Tyrosine codons" do
    assert ProteinTranslation.of_codon("UAU") == {:ok, "Tyrosine"}
    assert ProteinTranslation.of_codon("UAC") == {:ok, "Tyrosine"}
  end

  test "identifies Cysteine codons" do
    assert ProteinTranslation.of_codon("UGU") == {:ok, "Cysteine"}
    assert ProteinTranslation.of_codon("UGC") == {:ok, "Cysteine"}
  end

  test "identifies Tryptophan codons" do
    assert ProteinTranslation.of_codon("UGG") == {:ok, "Tryptophan"}
  end

  test "identifies stop codons" do
    assert ProteinTranslation.of_codon("UAA") == {:ok, "STOP"}
    assert ProteinTranslation.of_codon("UAG") == {:ok, "STOP"}
    assert ProteinTranslation.of_codon("UGA") == {:ok, "STOP"}
  end

  test "translates rna strand into correct protein" do
    strand = "AUGUUUUGG"
    assert ProteinTranslation.of_rna(strand) == {:ok, ~w(Methionine Phenylalanine Tryptophan)}
  end

  test "stops translation if stop codon present" do
    strand = "AUGUUUUAA"
    assert ProteinTranslation.of_rna(strand) == {:ok, ~w(Methionine Phenylalanine)}
  end

  test "stops translation of longer strand" do
    strand = "UGGUGUUAUUAAUGGUUU"
    assert ProteinTranslation.of_rna(strand) == {:ok, ~w(Tryptophan Cysteine Tyrosine)}
  end

  test "invalid RNA" do
    assert ProteinTranslation.of_rna("CARROT") == {:error, "invalid RNA"}
  end

  test "invalid codon at end of RNA" do
    assert ProteinTranslation.of_rna("UUUROT") == {:error, "invalid RNA"}
  end

  test "invalid codon" do
    assert ProteinTranslation.of_codon("INVALID") == {:error, "invalid codon"}
  end
end

File: typescript/protein-translation/protein-translation.test.ts
import { translate } from "./protein-translation";

describe("Translate input RNA sequences into proteins", () => {
  it("Methionine RNA sequence", () => {
    const expected = ["Methionine"];
    expect(translate("AUG")).toEqual(expected);
  });

  it("Phenylalanine RNA sequence 1", () => {
    const expected = ["Phenylalanine"];
    expect(translate("UUU")).toEqual(expected);
  });

  it("Phenylalanine RNA sequence 2", () => {
    const expected = ["Phenylalanine"];
    expect(translate("UUC")).toEqual(expected);
  });

  it("Leucine RNA sequence 1", () => {
    const expected = ["Leucine"];
    expect(translate("UUA")).toEqual(expected);
  });

  it("Leucine RNA sequence 2", () => {
    const expected = ["Leucine"];
    expect(translate("UUG")).toEqual(expected);
  });

  it("Serine RNA sequence 1", () => {
    const expected = ["Serine"];
    expect(translate("UCU")).toEqual(expected);
  });

  it("Serine RNA sequence 2", () => {
    const expected = ["Serine"];
    expect(translate("UCC")).toEqual(expected);
  });

  it("Serine RNA sequence 3", () => {
    const expected = ["Serine"];
    expect(translate("UCA")).toEqual(expected);
  });

  it("Serine RNA sequence 4", () => {
    const expected = ["Serine"];
    expect(translate("UCG")).toEqual(expected);
  });

  it("Tyrosine RNA sequence 1", () => {
    const expected = ["Tyrosine"];
    expect(translate("UAU")).toEqual(expected);
  });

  it("Tyrosine RNA sequence 2", () => {
    const expected = ["Tyrosine"];
    expect(translate("UAC")).toEqual(expected);
  });

  it("Cysteine RNA sequence 1", () => {
    const expected = ["Cysteine"];
    expect(translate("UGU")).toEqual(expected);
  });

  it("Cysteine RNA sequence 2", () => {
    const expected = ["Cysteine"];
    expect(translate("UGC")).toEqual(expected);
  });

  it("Tryptophan RNA sequence", () => {
    const expected = ["Tryptophan"];
    expect(translate("UGG")).toEqual(expected);
  });

  it("STOP codon RNA sequence 1", () => {
    const expected: string[] = [];
    expect(translate("UAA")).toEqual(expected);
  });

  it("STOP codon RNA sequence 2", () => {
    const expected: string[] = [];
    expect(translate("UAG")).toEqual(expected);
  });

  it("STOP codon RNA sequence 3", () => {
    const expected: string[] = [];
    expect(translate("UGA")).toEqual(expected);
  });

  it("Translate RNA strand into correct protein list", () => {
    const expected = ["Methionine", "Phenylalanine", "Tryptophan"];
    expect(translate("AUGUUUUGG")).toEqual(expected);
  });

  it("Translation stops if STOP codon at beginning of sequence", () => {
    const expected: string[] = [];
    expect(translate("UAGUGG")).toEqual(expected);
  });

  it("Translation stops if STOP codon at end of two-codon sequence", () => {
    const expected = ["Tryptophan"];
    expect(translate("UGGUAG")).toEqual(expected);
  });

  it("Translation stops if STOP codon at end of three-codon sequence", () => {
    const expected = ["Methionine", "Phenylalanine"];
    expect(translate("AUGUUUUAA")).toEqual(expected);
  });

  it("Translation stops if STOP codon in middle of three-codon sequence", () => {
    const expected = ["Tryptophan"];
    expect(translate("UGGUAGUGG")).toEqual(expected);
  });

  it("Translation stops if STOP codon in middle of six-codon sequence", () => {
    const expected = ["Tryptophan", "Cysteine", "Tyrosine"];
    expect(translate("UGGUGUUAUUAAUGGUUU")).toEqual(expected);
  });
});
